src/components/KeyboardAvoidingView/index.tsx

```diff
@@ -0,0 +1,145 @@
+import React, { forwardRef, useCallback } from 'react';
+import {
+  LayoutRectangle,
+  useWindowDimensions,
+  View,
+  ViewProps,
+} from 'react-native';
+import Reanimated, {
+  useAnimatedStyle,
+  useWorkletCallback,
+  useSharedValue,
+  useDerivedValue,
+  interpolate,
+} from 'react-native-reanimated';
+import { useKeyboardAnimation } from './hooks';
+
+type Props = {
+  /**
+   * Specify how to react to the presence of the keyboard.
+   */
+  behavior?: 'height' | 'position' | 'padding';
+
+  /**
+   * Style of the content container when `behavior` is 'position'.
+   */
+  contentContainerStyle?: ViewProps['style'];
+
+  /**
+   * Controls whether this `KeyboardAvoidingView` instance should take effect.
+   * This is useful when more than one is on the screen. Defaults to true.
+   */
+  enabled?: boolean;
+
+  /**
+   * Distance between the top of the user screen and the React Native view. This
+   * may be non-zero in some cases. Defaults to 0.
+   */
+  keyboardVerticalOffset?: number;
+} & ViewProps;
+
+const defaultLayout: LayoutRectangle = {
+  x: 0,
+  y: 0,
+  width: 0,
+  height: 0,
+};
+
+/**
+ * View that moves out of the way when the keyboard appears by automatically
+ * adjusting its height, position, or bottom padding.
+ */
+const KeyboardAvoidingView = forwardRef<View, React.PropsWithChildren<Props>>(
+  (
+    {
+      behavior,
+      children,
+      contentContainerStyle,
+      enabled = true,
+      keyboardVerticalOffset = 0,
+      style,
+      onLayout: onLayoutProps,
+      ...props
+    },
+    ref
+  ) => {
+    const initialFrame = useSharedValue<LayoutRectangle | null>(null);
+    const frame = useDerivedValue(() => initialFrame.value || defaultLayout);
+
+    const keyboard = useKeyboardAnimation();
+    const { height: screenHeight } = useWindowDimensions();
+
+    const relativeKeyboardHeight = useWorkletCallback(() => {
+      const keyboardY =
+        screenHeight - keyboard.heightWhenOpened.value - keyboardVerticalOffset;
+
+      return Math.max(frame.value.y + frame.value.height - keyboardY, 0);
+    }, [screenHeight, keyboardVerticalOffset]);
+
+    const onLayout = useCallback<NonNullable<ViewProps['onLayout']>>(
+      (e) => {
+        if (initialFrame.value === null) {
+          initialFrame.value = e.nativeEvent.layout;
+        }
```

Comment:
Could you please advise me what happens if I change the height of the container? I think initialFrame.value will have the initial values and it may cause some bugs. Maybe it would be better to recalculate them every change the size?

Reply:
@IvanIhnatsiuk yes, I know about it. The layout should be updated when keyboard is not shown. In current implementation if I remove `if`-statement above - this function will be called every frame when keyboard starts movement, and calculation will be messed up.

I'm going to improve and re-work this piece of code, but in other PR - the current implementation has some edge cases which I'm going to fix a little bit later and I'd like to have a chain of PRs for better tracing of history of changes 😊 (the current PR already big enough and in most of the cases it should work identically to RN variant).